### PR TITLE
2416 associade dissociate hts refactor (after #2516)

### PR DESCRIPTION
**Pull request by stoqnkpL**
This PR refactors the `AssociateTransitionLogic` and the `DissociateTransitionLogic`.
The refactor is aligned with the new pattern of introducing discrete logic classes, which hold the business logic in order to provide for interoperability in the context of standard Hedera operations and precompile operations.

Notes to the reviewer
The TokenAssociationSpecs is behaving inconsistently.

This PR is dependant on https://github.com/hashgraph/hedera-services/pull/2516

- [x] Passing E2E
- [x] Passing unit tests

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenAssociateTransitionLogic.java [hunk 1]
```diff
@@ -50,53 +50,25 @@
...
-	private final AccountStore accountStore;
-	private final TypedTokenStore tokenStore;
+
 	private final TransactionContext txnCtx;
-	private final GlobalDynamicProperties dynamicProperties;
+	private final AssociateLogic associateLogic;
 
+	// TODO: Global dynamic properties are not being used in the transition logic itself and is being moved to the
+	//  	 Associate logic. Confirm, that this solution is appropriate.
 	@Inject
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/txns/token/TokenAssociateTransitionLogic.java:59]: Seems fine to me, I think we can remove this TODO

File: hedera-node/src/main/java/com/hedera/services/txns/token/AssociateLogic.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package com.hedera.services.txns.token;
+
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/txns/token/AssociateLogic.java:2]: nit: license

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenDissociateTransitionLogic.java [hunk 3]
```diff
@@ -52,50 +44,24 @@
...
-			OptionValidator validator,
-			DissociationFactory dissociationFactory
-	) {
-		this.accountStore = accountStore;
-		this.tokenStore = tokenStore;
+			DissociateLogic dissociateLogic) {
 		this.txnCtx = txnCtx;
-		this.validator = validator;
-		this.dissociationFactory = dissociationFactory;
+
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/txns/token/TokenDissociateTransitionLogic.java:55]: nit: we can remove this empty line

File: hedera-node/src/main/java/com/hedera/services/store/models/NftId.java [hunk 4]
```diff
@@ -36,6 +36,13 @@ public NftId(long shard, long realm, long num, long serialNo) {
 		this.serialNo = serialNo;
 	}
 
+	public NftId(long num, long serialNo) {
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/store/models/NftId.java:39]: I think it's better to remove this constructor and not hardcode shard and realm to 0s. In the future those fields can have a different value and then it wouldn't be correct to assume they're zeros.
Instead, I would suggest to add a static method in EntityIdUtils: `asNftId(Id tokenId, long serialNumer)` which calls the existing NftId constructor with the realm and shard values of the Id.

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenAssociateTransitionLogic.java (hunk 1 above)
Comment (CordonaCodeCraft) [reply hedera-node/src/main/java/com/hedera/services/txns/token/TokenAssociateTransitionLogic.java:59]: Done.

File: hedera-node/src/main/java/com/hedera/services/txns/token/AssociateLogic.java (hunk 2 above)
Comment (CordonaCodeCraft) [reply hedera-node/src/main/java/com/hedera/services/txns/token/AssociateLogic.java:2]: Done

**Comment by tinker-michaelj** [comment]
> Notes to the reviewer
The `TokenAssociationSpecs` is behaving inconsistently.

I suspect this was caused by the [bug in `EntityNumPair.asTokenIdAndSerialPair()`](https://github.com/hashgraph/hedera-services/blob/3d03f0ac7e6523a177cf57e831903eafe29d2688/hedera-node/src/main/java/com/hedera/services/utils/EntityNumPair.java#L77) that used the serial number as _also_ the token id num. 😀

**Pull request merged**